### ISSUE TITLE
DeepAR: inaccurate forecasts on 5 minute data

### ISSUE THREAD

**Issue by Msaleh87** [awslabs/gluon-ts#2242]
### Discussed in https://github.com/awslabs/gluon-ts/discussions/2241

<div type='discussions-op-text'>

<sup>Originally posted by **Msaleh87** August 25, 2022</sup>
I have an electric load data for about a year, I trained the model for a data set from Mar 2021 until Mar 2022. I only used one feature (zeros and ones for weekends and weekdays). Here is my code:

```
estimator = DeepAREstimator(freq = '5min',
                           context_length = 7*24*12,  #The previous data points I look at to predict. It's like n_forecast in Neural Prophet 
                           prediction_length = 1*24*12,
                           num_layers = 2,
                           num_cells = 64,
                           cell_type = 'lstm',
                           use_feat_dynamic_real =True,
                           trainer = Trainer(epochs= 50))

from gluonts.dataset.common import ListDataset
training_data = ListDataset([{"start": df_input.index[0],
                              "target": df_input.Load[:'2022-03-07 04:50:00'],
                              'feat_dynamic_real': [df_input.DofW[:'2022-03-07 04:50:00']]}],
                           freq = "5min")

predictor = estimator.train(training_data = training_data)

test_data = ListDataset(
    [
        {"start": df_input.index[0], "target": df_input.Load[:'2022-03-08 04:50:00'], 'feat_dynamic_real': [df_input.DofW[:'2022-03-08 04:50:00']]}, # one day
        {"start": df_input.index[0], "target": df_input.Load[:'2022-03-09 04:50:00'], 'feat_dynamic_real': [df_input.DofW[:'2022-03-09 04:50:00']]}   # two days
    
    ],

freq = "5min")

forecast_it, ts_it = make_evaluation_predictions(
    dataset = test_data,
    predictor = predictor,
    num_samples = 1000 # I used 100 as well and it didn't make a big difference
)
```
These are the results I got, after almost two hours of training. What hyperparameters I should add/ tune to make the forecast better? I am hoping to use it to predict 48 hours ahead. Also, I used Neural Prophet and it gave me way better results for 48 hours ahead so far.

![Figure 1 (8)](https://user-images.githubusercontent.com/42920113/186670493-c587de50-5ea2-434a-ba65-5fd1d34d85d6.png)
![Figure 2 (4)](https://user-images.githubusercontent.com/42920113/186670521-99d83d8b-1ec3-4f2b-a8bc-250fcb763f7e.png)

</div>

**Comment by lostella** [comment]
@Msaleh87 thanks for opening the issue. It appears that for 5min data, by default, DeepAR looks only 1153 points in the past. This is about 4 days, so the model cannot capture the weekly seasonality properly. You should be able to get better accuracy by constructing the estimator as follows:

```python
from gluonts.time_feature import get_lags_for_frequency

estimator = DeepAREstimator(
    [...], # everything you have already
    lags_seq=get_lags_for_frequency("5min", lag_ub=2400),
)
```

The added `lags_seq` option will set the model up with a longer "receptive field", and should capture the seasonality much better.

Please let me know if this works! In any case, leaving this issue open since I believe the code should be fixed to work better on 5min data.

**Comment by lostella** [comment]
@Msaleh87 as a side note: I noticed you opened a few discussions and then turned them into issues. In general, it would be better to avoid posting twice since that may spread the discussion in two places, and potentially create confusion. As a rule of thumb (but sometimes the distinction is subtle) we should try to use
* discussions for questions and help requests ("how can I get X done?", "is my understanding of Y correct?", and so on),
* issues for discussing problems ("Z does not appear to be working"), detailed feature requests, other enhancement proposals (changes in APIs or similar)

**Comment by Msaleh87** [comment]
Sorry about that, but sometimes I don't get any reply in the discussion section and I think maybe there is a bigger issue and I need to put it as an issue. Because I usually do my homework and read most of the documentation before I ask, I used this for GluonTs: https://ts.gluon.ai/stable/tutorials/forecasting/extended_tutorial.html#Configuring-an-estimator
I will try to keep my questions to a minimum and be specific. 

Thanks for telling me about " It appears that for 5min data, by default, DeepAR looks only 1153 points in the past." However, whenever I increased the context_length, the model used to take longer to train! 

I added the part that you suggested to my code, and these are the results I got:

Code:
```
from gluonts.time_feature import get_lags_for_frequency
from gluonts.mx.trainer.learning_rate_scheduler import LearningRateReduction

scheduler = LearningRateReduction(patience=10, base_lr=0.001, objective='min')
estimator = DeepAREstimator(freq = '5min',
                           context_length = 7*24*12,  # <-------------
                           prediction_length = 2*24*12,
                           num_layers = 2,
                           num_cells = 128, 
                           #cardinality = 'auto',
                           scaling = True,
                           #use_feat_dynamic_Real =True, 
                           #early_stopping_patience = 6,
                           trainer = Trainer(epochs= 20, learning_rate = 0.001,hybridize=True, callbacks=[scheduler]),
                           lags_seq = get_lags_for_frequency("5min", lag_ub=7*24*12)) # <-------------

from gluonts.dataset.common import ListDataset

training_data = ListDataset([{"start": df_input.index[0],
                              "target": df_input.Load[:'2022-03-07 04:50:00'],
                              'feat_dynamic_cat': [df_input.DofW[:'2022-03-07 04:50:00']],
                              'time_feat': [time_feat_df.DayofWeek[:'2022-03-07 04:50:00'],
                                            time_feat_df.Hour[:'2022-03-07 04:50:00'],] }],
                           freq = "5min")
```
These are the results I got. Do you think I could get better results?
![Figure 5 (1)](https://user-images.githubusercontent.com/42920113/187766426-829cdf6f-46ab-4e52-9496-6504fde92c2f.png)
![Figure 6 (1)](https://user-images.githubusercontent.com/42920113/187766453-0997faa7-f4e1-4754-b7a6-496dd1d22a8b.png)

Thank you so much for your help!

**Comment by lostella** [comment]
@Msaleh87 I only now realized how large a `context_length` you are using: I would keep that to a much smaller value, say equal to `prediction_length`. No need to make it one week.

**Comment by Msaleh87** [comment]
@lostella, I am a bit confused now, what's the difference between context_length and lags_seq=get_lags_for_frequency("5min", lag_ub=2400)

**Comment by lostella** [comment]
@Msaleh87 `context_length` determines how many time steps before prediction the RNN unrolls; the lags indicate what lagged observations the RNN takes as input at each time step. In practice, the model will “look back” until `context_length + max(lags)` time steps before prediction.

You can try leaving `context_length` to its default.

**Comment by Msaleh87** [comment]
Thanks @lostella, I made the context_length = 4 days, and the lags = 14 days. I used this code:

```
learning_rate = 0.001
scheduler = LearningRateReduction(patience=25, base_lr=learning_rate, objective='min')

estimator = DeepAREstimator(freq = '5min',
                           context_length = 4*24*12,   # <------
                           prediction_length = 2*24*12,
                           num_layers = 2,
                           num_cells = 128, 
                           #cardinality = 'auto',
                           scaling = True,
                           #use_feat_dynamic_real =True,
                           trainer = Trainer(epochs= 50, learning_rate = 0.001,hybridize=True, callbacks=[scheduler]), # <------
                           lags_seq = get_lags_for_frequency("5min", lag_ub=14*24*12))  # <--------
```
and I got the results below, what do you think I could modify to capture the valleys better? As you can see, the peaks are somewhat captured. I am using the hour of the day and the day of the week as features, along with holidays and weekends.

![Figure 5 (1)](https://user-images.githubusercontent.com/42920113/188219004-40b65472-04ed-47f8-9b0e-9ef3866db38b.png)

![Figure 6 (1)](https://user-images.githubusercontent.com/42920113/188219054-998e19b7-dddb-4ec9-875b-bde7175191eb.png)

**Comment by Msaleh87** [comment]
@lostella Thanks, I am a bit confused about the time_feat, do I need to enable it inside the estimator? I ran the predictor and it runs fine, but I am not sure if it's using the time_feat I added to the training data set. I added 0 for weekends and 1 for weekdays as  'feat_dynamic_cat'. Also, I added in the time_feat, the hour of the day, and the day of the week (do I need to add those, or the model picks up on this?)

Here is my code:
```
estimator = DeepAREstimator(freq = '15min',
                           context_length = 4*24*4, 
                           prediction_length = 2*24*4,
                           num_layers = 2,
                           num_cells = 128,
                           scaling = True,
                           trainer = Trainer(epochs= 20, learning_rate = 0.001,hybridize=True, callbacks=[scheduler]),
                           lags_seq = get_lags_for_frequency("15min", lag_ub=14*24*4)) 

from gluonts.dataset.common import ListDataset

training_data = ListDataset([{"start": df_input.index[0],
                              "target": df_input.Load[:'2022-03-07 04:45:00'],
                              'feat_dynamic_cat': [df_input.DofW[:'2022-03-07 04:45:00']], # 0 if weekend, 1 if weekday
                              'time_feat': [time_feat_df.DayofWeek[:'2022-03-07 04:45:00'], #1 for Monday, 2 for Tuesday, and so on
                                            time_feat_df.Hour[:'2022-03-07 04:45:00'],] }], # the hour of the day 1-->24
                           freq = "15min")
```

**Comment by khawar-islam** [comment]
@Msaleh87 Can you please help me? I am facing issues while training and my training is extremely slow and I have no idea why? do you think is it correct?

```
# Define the DeepAR estimator
predictor = DeepAREstimator(
    prediction_length=12,  # Adjust based on how far you want to predict
    context_length=24,  # Context length should be at least as long as prediction length
    freq="1min",  # Change to your data's frequency
    batch_size=64,
    trainer_kwargs={"max_epochs": 1, "accelerator": "gpu"}
).train(training_data, )
```
